lib/posix/clock.c

```diff
@@ -106,6 +105,66 @@ int clock_settime(clockid_t clock_id, const struct timespec *tp)
 	return 0;
 }
 
+/**
+ * @brief Suspend execution for a nanosecond interval, or
+ * until some absolute time relative to the specified clock.
+ *
+ * See IEEE 1003.1
+ */
+int clock_nanosleep(clockid_t clock_id, int flags, const struct timespec *rqtp,
+		    struct timespec *rmtp)
+{
+	uint64_t ns;
+	uint64_t us;
+	uint64_t uptime_ns;
+	k_spinlock_key_t key;
+	const bool update_rmtp = rmtp != NULL;
+
+	if (clock_id != CLOCK_REALTIME && clock_id != CLOCK_MONOTONIC) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if ((flags & TIMER_ABSTIME) != TIMER_ABSTIME) {
+		return nanosleep(rqtp, rmtp);
```

Comment:
In most implementations I would guess that `nanosleep(rqtp, rmtp)` is probably shorthand for `clock_nanosleep(CLOCK_MONOTONIC, 0, rqtp, rmtp)`.

But sticking with this for now is ok.

I sort of wonder, wouldn't it be easier to cover the absolute time to a relative time first and then just make `nanosleep()` do the heavy lifting?

Reply:
Yeh, as you say `nanosleep` is a special case of `clock_nanosleep`. So I think it would make sense for `nanosleep` to just `return clock_nanosleep(CLOCK_.REALTIME, 0, rqtp, rmtp)`.

I would need to change `clock_nanosleep` to handle both relative and absolute times. But would cut down on a lot of redundancy that exists between `nanosleep` and `clock_nanosleep` at the moment.

Converting an absolute timeout into a relative one would look something like:

1. call `clock_gettime()`
2. compute difference between absolute time and current time.
3. call `nanosleep` with a delay of this difference. 

If we are preempted between 1 and 3, then we will sleep for longer than wanted. An absolute sleep avoids this, issue and that is why I feel we should leave it in.

Reply:
Could we not instead convert relative time to absolute time in clock_nanosleep?

Reply:
Of course - whatever you are more comfortable with 👍